hazelcast/src/main/java/com/hazelcast/nio/tcp/WriteHandler.java

```diff
@@ -172,7 +172,7 @@ private void unschedule() {
         // So the outputBuffer is empty, so we are going to unschedule ourselves.
         scheduled.set(false);
 
-        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty()) {
+        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty() && currentPacket == null) {
```

Comment:
Peter, refactoring code; making it more simpler and more readable is good.
But what I get from Sancar's statement; this caused several critical bugs and time spent to solve them.

Possible reasons? 
- Mistake of implementor or reviewers (this is always possible)
- Low coverage of tests that touches such core parts (how possible is it to test such low level functionality)

I know you are planning to refactor/cleanup other parts of the code:
So what should be done to prevent such problems in future?